### PR TITLE
Refactor flow engine

### PR DESCRIPTION
**Pull request by jlowin**
Mirror of #13793, refactoring the "caller" functions of the flow engine to remove the need for them to know how the engine works and stay as lightweight as possible.

### CODE REVIEW

File: src/prefect/flow_engine.py [hunk 1]
```diff
@@ -382,45 +381,27 @@ def get_hooks(self, state: State, as_async: bool = False) -> Iterable[Callable]:
             hooks = flow.on_crashed_hooks
         elif state.is_running() and flow.on_running_hooks:
             hooks = flow.on_running_hooks
+        else:
+            hooks = None
 
         for hook in hooks or []:
```
Comment (chrisguidry) [inline src/prefect/flow_engine.py:387]:
```suggestion
        else:
            hooks = []

        for hook in hooks:
```

Maybe like 0.3% cleaner?

File: src/prefect/flow_engine.py [hunk 2]
```diff
@@ -569,61 +553,80 @@ def is_pending(self) -> bool:
...
-                except TimeoutError as exc:
-                    run.handle_timeout(exc)
-                except Exception as exc:
-                    # If the flow fails, and we have retries left, set the flow to retrying.
-                    run.logger.exception("Encountered exception during execution:")
-                    run.handle_exception(exc)
+    with engine.start():
+        while engine.is_running():
+            with engine.run_context():
+                engine.call_flow_fn()
```
Comment (chrisguidry) [inline src/prefect/flow_engine.py:624]: :heart:

**Review by chrisguidry — APPROVED** [review]
Make sense to me!  Very nice reduction

File: src/prefect/flow_engine.py (hunk 1 above)
Comment (jlowin) [reply src/prefect/flow_engine.py:387]: every % counts

**Pull request merged**